docs/reference/apply_and_sync.md

```diff
@@ -0,0 +1,88 @@
+
+This is because in the database everything is mutable. But the cloud is not exactly like that. Some cloud entities cannot be changed after they have been created. They can only be replaced. You also can't just issue a delete and then an insert, because other entities that depend on the entity you're trying to replace are still there and the delete operation on your current entity will not eliminate those other entities, but will instead error out.
+
+What you need to do is only create a new entity that matches the database, and then on the next loop it will notice that it should delete the entity that isn't in the database and the *other* entities will notice that they are associated with the wrong entity and trigger their own update or replace to swap to the new entity you just created. These tasks will potentially clash with each other but the error-retrying-loop should eventually execute them in the correct order and you're done.
+
+That explains one loop, so why the other? Because of create-only entities that *also* have unique constraints on one or more properties. The AWS Security Group qualifies as one of these. Once you create a security group and set its description, you can't change it, but in IaSQL you certain can do so. If you update the description, which is not unique, but leave the name alone, which is unique, you can't just create a replacement security group with the same name. So the security group mapper has to do something gnarly here. When it creates the replacement, it mutates the name to a randomly generated one and then mutates the internal cache of the DB record to match that randomly generated name.
+
+Now when it does this replace, it first sees that it is supposed to mutate the security group description without mutating the name but alters the name in its cache and executes a cloud create with this new, fake record. Then the inner loop runs and entities connected to the security group "realize" they are on the wrong security group and switch to this new, temporary security group, while the old security group with the correct name is marked as cloud-only so it is deleted. Then the inner loop completes with no difference between the database *cache* and the cloud, so it goes back to the outer loop, which re-loads the *actual* database state and it checks again and sees that there's a security group in the database that is not in the cloud and a security group in the cloud that is not in the database, and that these other entities are connected to the wrong security group, so it creates tasks to create a security group, delete a security group, and switch security groups. The error-retrying-loop will execute these tasks some unknown number of times, but the successes will always be in the same order -> create security group -> switch other entities to that security group -> delete the temporary security group.
```

Comment:
This loop will be hard to follow for someone coming it. We have the security group example, but it might be a good idea to re-state at the end what the user does exactly (SQL included?), what the loop does under the hood and what end/goal state is achieved?